lib/src/input_validation/validate_function.dart

```diff
@@ -1,103 +1,55 @@
 import 'package:advanced_calculation/src/input_validation/error_state.dart';
-import 'package:advanced_calculation/src/input_validation/start_state.dart';
-import 'package:advanced_calculation/src/input_validation/state.dart';
-import 'open_subexpression_state.dart';
 import 'package:advanced_calculation/src/input_validation/pattern.dart';
+import 'package:advanced_calculation/src/input_validation/tracking/validation_tracking.dart';
+
+import 'state.dart';
 
 class ValidateFunction {
-  State currentState;
-  bool isMultiParam = false;
-  int multiParamCounter = 0; // Considers counting '(' for math functions with more than parameter
-  int counter = 0; // Considers counting '(' for math functions with 1 parameter
-  static final List<String> lengthTwoFunc = ["ln"];
-  static final List<String> lengthThreeFunc = ["log","sin","cos","tan", "abs", "csc","sec", "cot" ];
-  static final List<String> lengthFourFunc = ["sqrt", "sinh", "cosh", "tanh", "asin", "acos", "atan", "acsc", "asec", "acot", "csch", "sech", "coth", "ceil"];
-  static final List<String> lengthFiveFunc = ["asinh", "acosh", "atanh", "acsch", "asech", "acoth", "floor", "round", "trunc", "fract"];
-  static final List<String> lengthThreeMultiParamFunc = ["max", "min", "gcd", "lcm"];
+  static final List<String> validFunctions = ["ln","log","sin","cos","tan", "abs", "csc","sec", "cot", "sqrt", "sinh", "cosh", "tanh",
+    "asin", "acos", "atan", "acsc", "asec", "acot", "csch", "sech", "coth", "ceil","asinh", "acosh", "atanh", "acsch", "asech",
+    "acoth", "floor", "round", "trunc", "fract"];
+  static final List<String> multiParamFunctions = ["max", "min", "gcd", "lcm"];
 
-  ValidateFunction(){
-    currentState= new StartState(this);
-  }
 
-  void setCurrentState(State currentState) {
-    this.currentState = currentState;
-  }
+  List<String> _sanitizeInput(String input){
+    List<String> sanitizedInput = input.split(" ").where((item) => item.isNotEmpty).toList();
+    sanitizedInput.add('=');
 
-  State getCurrentState() {
-    return currentState;
+    return sanitizedInput;
   }
 
-  List<String> initialize(String input){
-    this.counter = 0;
-    this.multiParamCounter = 0;
-    isMultiParam = false;
-    input = input + " = ";
-    List<String> inputString = input.split(" ");
-
-    for(int i = 0; i < inputString.length; i++){
-      if(inputString[i].isEmpty){
-        inputString.removeAt(i);
-      }
+  String _sanitizeToken(String token){
+    String sanitizedToken = token;
+    //handle special negatives for complex functions
+    if(sanitizedToken.startsWith('-') && sanitizedToken.length > 1) {
+      sanitizedToken = sanitizedToken.substring(1); // remove the negative
```

Comment:
this was supposed to be removed, it's handled by the translator

Reply:
definitely would like it removed, but would prefer to have this PR just be the cleanup